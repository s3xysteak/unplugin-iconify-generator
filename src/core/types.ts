export interface PluginOptions {
  /**
   * Support `Iconify IntelliSense`. ! The options will change your `.vscode/settings.json`.
   * Use string to specify the path of the settings.json file.
   * @default true
   */
  iconifyIntelliSense?: boolean | string

  /**
   * @default process.cwd()
   */
  base?: string

  /**
   * A key-value pair of icons collection names and the path of icons' folder.
   *
   * Key is the name of a collection.
   *
   * Value is the path of the folder where the icons are stored.
   */
  collections?: Record<string, string>

  /**
   * Indicates the output path of the iconify meta file.
   *
   * @default './node_modules/.unplugin-iconify-generator'
   */
  output?: string
}

export interface IconifyIcon {
  body: string

  // Left position of viewBox.
  // Defaults to 0.
  left?: number

  // Top position of viewBox.
  // Defaults to 0.
  top?: number

  // Width of viewBox.
  // Defaults to 16.
  width?: number

  // Height of viewBox.
  // Defaults to 16.
  height?: number
}

export interface IconifyJSONIconsData {
  // Prefix for icons in JSON file, required.
  prefix: string

  // List of icons, required.
  icons: Record<string, IconifyIcon>
}
