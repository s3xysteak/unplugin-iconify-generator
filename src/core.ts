export * from './core/parser'
export * from './core/types'
