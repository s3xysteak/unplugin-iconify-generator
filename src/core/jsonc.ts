import * as jsonc from 'jsonc-parser'

export function injectJsonc(jsonText: string, key: string, newValue: any, flat = true) {
  const { insertSpaces, tabSize } = detectIndentation(jsonText)

  const edits = jsonc.modify(
    jsonText,
    flat ? [key] : key.split('.'),
    newValue,
    { formattingOptions: { insertSpaces, tabSize } },
  )

  const updatedJsonText = jsonc.applyEdits(jsonText, edits)

  return updatedJsonText
}

export function detectIndentation(jsonText: string) {
  const lines = jsonText.split(/\r?\n/)
  for (const line of lines) {
    const match = line.match(/^(\s+)\S/)
    if (match) {
      const indent = match[1]
      return {
        insertSpaces: indent.includes(' '),
        tabSize: indent.includes(' ') ? indent.length : 2,
      }
    }
  }
  return {
    insertSpaces: true,
    tabSize: 2,
  }
}
