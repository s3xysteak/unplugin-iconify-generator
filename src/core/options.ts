import type { PluginOptions } from './types'
import process from 'node:process'
import { isAbsolute, normalize, resolve } from 'pathe'

// _internals
export function resolveOptions(userOptions: Partial<PluginOptions>): PluginOptions {
  const defaultOptions: PluginOptions = {
    cwd: process.cwd(),
    iconifyIntelliSense: true,
    output: './node_modules/.unplugin-iconify-generator',
    collections: {},
  }

  const result = {
    ...defaultOptions,
    ...userOptions,
  }

  return {
    ...result,
    output: isAbsolute(result.output) ? normalize(result.output) : resolve(result.cwd, result.output),
    collections: Object.fromEntries(
      Object.entries(result.collections)
        .map(([prefix, path]) => [
          prefix,
          (isAbsolute(path)
            ? normalize(path)
            : resolve(result.cwd, path)
          ).replace(/\/$/, ''),
        ]),
    ),
  }
}
