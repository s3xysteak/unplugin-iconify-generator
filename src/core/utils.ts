export function toArray<T>(array?: T | T[]): T[] {
  array = array ?? []
  return Array.isArray(array) ? array : [array]
}
