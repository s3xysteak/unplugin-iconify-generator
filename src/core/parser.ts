import type { IconifyIcon, IconifyJSONIconsData, PluginOptions } from './types'
import fs from 'fs-extra'
import { basename, extname, join, resolve } from 'pathe'
import { glob } from 'tinyglobby'
import { notNullish, objectMap } from './utils'

export async function normalizeIcons(options: PluginOptions, base: string): Promise<IconifyJSONIconsData[] | false> {
  return options.collections && Object.keys(options.collections).length > 0
    ? await Promise.all(
      Object.entries(options.collections).map(async ([prefix, path]) => {
        const paths = await glob([join(path, '*.svg')], { cwd: base })
          .then(relatives => relatives.map(relative => resolve(base, relative)))

        const icons = Object.fromEntries(
          await Promise.all(
            paths.map(async (path) => {
              const name = path.split('/').at(-1)
              if (!name)
                return

              const svg = await fs.readFile(path, 'utf-8')

              const iconifyIcon = parseIcon(svg)
              if (!iconifyIcon)
                return

              return [
                basename(name, extname(name)),
                iconifyIcon,
              ] as const
            }),
          ).then(v => v.filter(notNullish)),
        )

        return {
          prefix,
          icons,
        }
      }),
    )
    : false
}

/**
 * 1. width
 * 2. height
 * 3. body
 */
export const SVG_REGEX = /<svg[^>]*?viewBox="\s*\d+\s+\d+\s+(\d+)\s+(\d+)\s*"[^>]*>(.*?)<\/svg>/s

export function parseIcon(svg: string): IconifyIcon | null {
  const match = svg.match(SVG_REGEX)

  const width = Number(match?.[1])
  const height = Number(match?.[2])

  return !!match && !Number.isNaN(width) && !Number.isNaN(height)
    ? {
        width,
        height,
        body: match[3].trim(),
      }
    : null
}

export function parseIcons(icons: Record<string, string>): IconifyJSONIconsData['icons'] {
  return objectMap(icons, (k, v) => {
    const val = parseIcon(v)

    return val ? [k, val] : undefined
  })
}
