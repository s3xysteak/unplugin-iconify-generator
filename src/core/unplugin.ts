import type { IconifyJSONIconsData } from '@iconify/types'
import type { FSWatcher } from 'chokidar'
import type { PluginOptions } from './types'
import chokidar from 'chokidar'
import fs from 'fs-extra'
import { dirname, relative, resolve } from 'pathe'
import { debounce } from 'perfect-debounce'

import { createUnplugin } from 'unplugin'
import { writeIntoVscodeSettings } from './jsonc'
import { resolveOptions } from './options'
import { normalizeIcon } from './parser'
import { mapReverse } from './utils'

export function createContext(opts: PluginOptions) {
  /** output is absolute path */
  const prefixOutputMap = new Map(
    Object.keys(opts.collections)
      .map(prefix => [
        prefix,
        resolve(opts.output, `${prefix}.json`),
      ]),
  )

  /** prefix - update debounce */
  const callbacks = new Map<string, () => Promise<void>>()
  /** icons path - prefix[] */
  const iconPrefixesMap = mapReverse(new Map(Object.entries(opts.collections)))
  function getPrefixes(icon: string) {
    let iconDir: string | undefined

    const rel = dirname(icon)
    const abs = resolve(opts.cwd, rel)

    const prefixes = [
      ...iconPrefixesMap.has(rel) ? iconPrefixesMap.get(iconDir = rel)! : [],
      ...iconPrefixesMap.has(abs) ? iconPrefixesMap.get(iconDir = abs)! : [],
    ]

    return {
      prefixes,
      iconDir,
    }
  }

  const watchCb = async (icon: string) => {
    if (!icon.endsWith('.svg'))
      return

    const { iconDir, prefixes } = getPrefixes(icon)
    if (!prefixes || !iconDir)
      return warn(`Cannot find ${prefixes}: ${icon}`)

    await Promise.all(
      prefixes.map(async (prefix) => {
        if (callbacks.has(prefix)) {
          callbacks.get(prefix)!()
        }
        else {
          const cb = debounce(async () => {
            const output = prefixOutputMap.get(prefix)
            if (!output)
              return warn(`Cannot find ${prefix}: ${output}`)

            const iconify = await normalizeIcon(opts, prefix, iconDir)
            await fs.outputFile(output, JSON.stringify(iconify))
          }, 100)
          callbacks.set(prefix, cb)
          await cb()
        }
      }),
    )
  }

  let watcher: FSWatcher

  return {
    watchStart() {
      watcher = chokidar
        .watch(Object.values(opts.collections), {
          cwd: opts.cwd,
          persistent: true,
        })
        .on('add', watchCb)
        .on('change', watchCb)
        .on('unlink', watchCb)
    },
    async watchEnd() {
      await watcher.close()
    },
    async clearOutput() {
      await fs.pathExists(opts.output)
        .then(async is => is && await fs.emptyDir(opts.output))
    },
    async createEmptyIconifyJSON() {
      for (const [prefix, output] of prefixOutputMap.entries()) {
        await fs.outputFile(output, JSON.stringify(<IconifyJSONIconsData>{ prefix, icons: {} }))
      }
    },
    async writeVscodeSettings() {
      if (opts.iconifyIntelliSense)
        await writeIntoVscodeSettings(opts, Array.from(prefixOutputMap.values()).map(p => relative(opts.cwd, p)))
    },
  }
}

export default createUnplugin<Partial<PluginOptions> | undefined>((userOptions = {}) => {
  const opts = resolveOptions(userOptions)

  const ctx = createContext(opts)

  return {
    name: 'unplugin-iconify-generator',
    async buildStart() {
      await ctx.clearOutput()

      await ctx.createEmptyIconifyJSON()

      await ctx.writeVscodeSettings()

      ctx.watchStart()
    },
    async buildEnd() {
      await ctx.watchEnd()
    },
  }
})

function warn(msg: string) {
  return console.warn(`[iconify] ${msg}`)
}
