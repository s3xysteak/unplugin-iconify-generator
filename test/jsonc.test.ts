import { describe, expect, it } from 'vitest'
import { detectIndentation, injectJsonc } from '../src/core/jsonc'

describe('jsonc', () => {
  it('detectIndentation', () => {
    expect(
      detectIndentation(`{
  "a": 1
}`),
    ).toEqual({ insertSpaces: true, tabSize: 2 })

    expect(
      detectIndentation(`{
    "a": 1
}`),
    ).toEqual({ insertSpaces: true, tabSize: 4 })
  })

  it('injectJsonc', () => {
    expect(injectJsonc(`{
  // hello
  "one": 1,
  // world
  "two": {
    three: 3 // UwU
  },
  "four": 4
}`, 'four', [4, 5, 6])).toMatchSnapshot()

    expect(injectJsonc(`{
  // hello
  "one": 1, // UwU
}`, 'two', 2)).toMatchSnapshot()

    expect(injectJsonc(`{
  // hello
  "one": 1, // UwU
  "two": [0, 0]
}`, 'two', [2, 3, 4])).toMatchSnapshot()

    expect(injectJsonc(`{
  // hello
  "one": 1,
}`, 'two.three', [2, 3])).toMatchSnapshot()
  })
})
