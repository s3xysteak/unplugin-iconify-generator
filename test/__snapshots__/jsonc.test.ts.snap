// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsonc > injectJsonc 1`] = `
"{
  // hello
  "one": 1,
  // world
  "two": {
    three: 3 // UwU
  },
  "four": [
    4,
    5,
    6
  ]
}"
`;

exports[`jsonc > injectJsonc 2`] = `
"{
  // hello
  "one": 1,
  "two": 2, // UwU
}"
`;

exports[`jsonc > injectJsonc 3`] = `
"{
  // hello
  "one": 1, // UwU
  "two": [
    2,
    3,
    4
  ]
}"
`;

exports[`jsonc > injectJsonc 4`] = `
"{
  // hello
  "one": 1,
  "two.three": [
    2,
    3
  ],
}"
`;
